proc-macros/src/new/lifetimes.rs

```diff
@@ -0,0 +1,118 @@
+use proc_macro2::Span;
+use syn::spanned::Spanned;
+use syn::{Lifetime, Path, Type};
+
+static LIFETIME: &str = "'a";
+
+pub fn replace_lifetimes(ty: &mut Type) -> bool {
+	let mut replaced = false;
+	traverse_type(ty, &mut replaced, replace_lifetime);
+	replaced
+}
+
+pub fn replace_lifetime(ty: &mut Type, replaced: &mut bool) {
+	use syn::{GenericArgument, PathArguments};
+
+	match ty {
+		Type::Path(p) => {
+			for segment in p.path.segments.iter_mut() {
+				if let PathArguments::AngleBracketed(ref mut ab) = segment.arguments {
+					let mut lifetimes = 0;
+
+					for arg in ab.args.iter_mut() {
+						match arg {
+							GenericArgument::Lifetime(lt) => {
+								*lt = Lifetime::new(LIFETIME, lt.span());
+								*replaced = true;
+								lifetimes += 1;
+							}
+							// Stop iterating on first non-lifetime generic argument
+							// TODO: Replace by `.iter_mut().map_while(...).count()` when it's stabilized
+							_ => break,
+						}
+					}
+
+					// Check if the type is a `Cow<str>` or similar with lifetime elision,
```

Comment:
Nitpick: "or similar"? we only check for exactly Cow, so I feel like those words can go :D

Reply:
I think we can drop it altogether, I wanted to be able to handle `Cow<str>`, but compiler throws an error if you do that regardless and requires you put in at least `Cow<'_, str>` :).

Reply:
Ah interesting!

Reply:
Keeping it in for now, I'll see if we can get away from using a trait so that we can actually accept `Cow<str>` in params :).